test/bm25_transformer.jl

```diff
@@ -0,0 +1,76 @@
+using MLJBase
+using TextAnalysis
+
+@testset "bm25 transformer" begin
+    # add some test docs
+    docs = ["Hi my name is Sam.", "How are you today?"]
+
+    # convert to ngrams
+    ngram_vec = ngrams.(documents(Corpus(NGramDocument.(docs))))
+
+    # train transformer
+    bm25_transformer = MLJText.BM25Transformer()
+    test_machine = @test_logs machine(bm25_transformer, ngram_vec)
+    MLJBase.fit!(test_machine)
+
+    # test
+    test_doc = ngrams(NGramDocument("Another sentence ok"))
+    test1 = transform(test_machine, [test_doc])
+    @test sum(test1, dims=2)[1] == 0.0
+    @test size(test1) == (1, 11)
+
+    test_doc2 = ngrams(NGramDocument("Listen Sam, today is not the day."))
+    test2 = transform(test_machine, [test_doc2])
+    @test sum(test2, dims=2)[1] > 0.0
+    @test size(test2) == (1, 11)
+
+    test_doc3 = ngrams.(
+        Corpus([NGramDocument("Another sentence ok"), NGramDocument("Listen Sam, today is not the day.")])
+    )
+    test3 = transform(test_machine, test_doc3)
+    @test sum(test3, dims=2)[1] == 0.0
+    @test sum(test3, dims=2)[2] > 0.0
+    @test size(test3) == (2, 11)
+
+    test_doc4 = [["Another", "sentence", "ok"], ["Listen", "Sam", ",", "today", "is", "not", "the", "day", "."]]
+    test4 = transform(test_machine, test_doc4)
+    @test sum(test4, dims=2)[1] == 0.0
+    @test sum(test4, dims=2)[2] > 0.0
+    @test size(test4) == (2, 11)
+    # test with bag of words
+    bag_of_words = Dict(
+        "cat in" => 1,
+        "the hat" => 1,
+        "the" => 2,
+        "cat" => 1,
+        "hat" => 1,
+        "in the" => 1,
+        "in" => 1,
+        "the cat" => 1
+    )
+    bag = Dict(Tuple(String.(split(k))) => v for (k, v) in bag_of_words)
+    bm25_transformer2 = MLJText.BM25Transformer()
+    test_machine2 = @test_logs machine(bm25_transformer2, [bag])
+    MLJBase.fit!(test_machine2)
+
+    test_doc5 = ["How about a cat in a hat"]
+    test5 = transform(test_machine2, test_doc5)
+    @test sum(test5, dims=2)[1] > 0.0
+    @test size(test5) == (1, 8)
+
+    docs = [
+        "the BIL opens the door to new possibilities and should raise our collective expectations",
+        "about what we can achieve in the near term.",
+        "the following projects are not yet at a stage where they could be competitive",
+        "for construction dollars over the next five years,",
+        "but with some attention and preliminary work, our transportation leaders could turn",
+        "these stretch projects into shovel-worthy ones."
+    ]
+    ngram_vec = ngrams.(documents(Corpus(NGramDocument.(docs))))
+    bm25_transformer3 = MLJText.BM25Transformer(max_doc_freq=0.8, min_doc_freq=0.2)
+    test_machine3 = @test_logs machine(bm25_transformer3, ngram_vec)
+    MLJBase.fit!(test_machine3)
+
+    test6 = transform(test_machine3, ngram_vec)
+    @test (Vector(vec(sum(test6, dims=2))) .> 0.8) == Bool[1, 1, 1, 1, 1, 1]
+end
```

Comment:
Just a suggestion - feel free to ignore. It seems the two `AbstractTextTransformers` share some behaviour that could be tested simultaneously to avoid code duplication and make maintenance easier. Maybe you could wrap those tests separately, as in 

```julia
for model in [:TfidfTransformer, :B25Transformer]
    if model == :TfidfTransformer
        some_result = 42
    else
        some_result = 3
    end
   quote
        @test some_property($model) == $some_result
        ...
    end |> eval
end
```
or some variation of this.

Reply:
Yeah - I want to try to consolidate the logic for these tests.  I'm just trying to figure out the best way to do so.

Reply:
I consolidated the tests - if there is a better way using more of the meta-programming approach you referenced above, can you point me to an example?  That is probably the area with Julia where I am the weakest.